lib/connections/limiter.go

```diff
@@ -12,44 +12,130 @@ import (
 }
 
-func (lim *limiter) newReadLimiter(r io.Reader, isLAN bool) io.Reader {
-	return &limitedReader{reader: r, limiter: lim, isLAN: isLAN}
+// This function sets limiters according to corresponding DeviceConfiguration
+func setLimitsForDevice(v config.DeviceConfiguration, readLimiter, writeLimiter *rate.Limiter) {
+	// The rate variables are in KiB/s in the config (despite the camel casing
+	// of the name). We multiply by 1024 to get bytes/s.
+	if v.MaxRecvKbps <= 0 {
+		readLimiter.SetLimit(rate.Inf)
+	} else {
+		readLimiter.SetLimit(1024 * rate.Limit(v.MaxRecvKbps))
+	}
+
+	if v.MaxSendKbps <= 0 {
+		writeLimiter.SetLimit(rate.Inf)
+	} else {
+		writeLimiter.SetLimit(1024 * rate.Limit(v.MaxSendKbps))
+	}
 }
 
-func (lim *limiter) newWriteLimiter(w io.Writer, isLAN bool) io.Writer {
-	return &limitedWriter{writer: w, limiter: lim, isLAN: isLAN}
+// This function handles removing, adding and updating of device limiters.
+// Pass pointer to avoid copying. Pointer already points to copy of configuration
+// so we don't have to worry about modifying real config.
+func (lim *limiter) processDevicesConfiguration(from, to config.Configuration) {
+	seen := make(map[protocol.DeviceID]struct{})
+
+	// Mark devices which should not be removed, create new limiters if needed and assign new limiter rate
+	for _, v := range to.Devices {
+		if v.DeviceID == to.MyID {
+			// This limiter was created for local device. Should skip this device
+			continue
+		}
+		seen[v.DeviceID] = struct{}{}
+
+		readLimiter, okR := lim.getOrCreateDeviceReadLimiter(v.DeviceID)
+		writeLimiter, okW := lim.getOrCreateDeviceWriteLimiter(v.DeviceID)
+
+		// limiters for this device are created so we can store previous rates for logging
+		previousReadLimit := readLimiter.Limit()
+		previousWriteLimit := writeLimiter.Limit()
+		currentReadLimit := rate.Limit(v.MaxRecvKbps) * 1024
+		currentWriteLimit := rate.Limit(v.MaxSendKbps) * 1024
+		if v.MaxSendKbps <= 0 {
+			currentReadLimit = rate.Inf
+		}
+		if v.MaxRecvKbps <= 0 {
+			currentWriteLimit = rate.Inf
+		}
+
+		// Nothing about this device has changed. Start processing next device
+		if okR && okW && currentReadLimit == previousReadLimit && currentWriteLimit == previousWriteLimit {
+			continue
+		}
+		setLimitsForDevice(v, readLimiter, writeLimiter)
+
+		readLimitStr := "is unlimited"
+		if v.MaxRecvKbps > 0 {
+			readLimitStr = fmt.Sprintf("limit is %d KiB/s", v.MaxRecvKbps)
+		}
+		writeLimitStr := "is unlimited"
+		if v.MaxSendKbps > 0 {
+			writeLimitStr = fmt.Sprintf("limit is %d KiB/s", v.MaxSendKbps)
+		}
+
+		l.Infof("Device %s send rate %s, receive rate %s", v.DeviceID, readLimitStr, writeLimitStr)
+	}
+
+	// Delete remote devices which were removed in new configuration
+	for _, v := range from.Devices {
+		if _, ok := seen[v.DeviceID]; !ok {
+			l.Debugf("deviceID: %s should be removed", v.DeviceID)
+			lim.deleteDeviceLimiters(v.DeviceID)
+		}
+	}
+
+	l.Debugln("Rebuild of device limiters map finished")
```

Comment:
This is a lie, as I don't think we're rebuilding stuff here.

Reply:
I can imagine stuff getting out of sync a bit if two saves happen in parallel, but to be honest, I am not worried about a case like that.

Reply:
That's why I added mutex at the beginning of CommitConfiguration call